lib/Conversion/SCFToCalyx/SCFToCalyx.cpp

```diff
@@ -1712,6 +1756,94 @@ class SCFToCalyxPass : public SCFToCalyxBase<SCFToCalyxPass> {
 private:
   LogicalResult partialPatternRes;
   std::shared_ptr<calyx::CalyxLoweringState> loweringState = nullptr;
+
+  FuncOp createNewTopLevelFcn(ModuleOp moduleOp, std::string &baseName) {
+    std::string newName = baseName;
+    unsigned counter = 0;
+    while (SymbolTable::lookupSymbolIn(moduleOp, newName)) {
+      newName = baseName + "_" + std::to_string(++counter);
+    }
+
+    OpBuilder builder(moduleOp.getContext());
+    builder.setInsertionPointToStart(moduleOp.getBody());
+
+    FunctionType funcType = builder.getFunctionType({}, {});
+
+    if (auto newFunc =
+            builder.create<FuncOp>(moduleOp.getLoc(), newName, funcType)) {
+      baseName = newName;
```

Comment:
Instead of passing a mutable parameter `baseName` and updating it, can you just get the name from the returned `FuncOp`?

Reply:
We need to change it in-place so that after setting the top-level function [here](https://github.com/llvm/circt/blob/db509755c08048ecac1f740ed06e4ed6f2e88e5a/lib/Conversion/SCFToCalyx/SCFToCalyx.cpp#L1855), we can directly use `topLevelFunction` [in `labelEntryPoint`](https://github.com/llvm/circt/blob/db509755c08048ecac1f740ed06e4ed6f2e88e5a/lib/Conversion/SCFToCalyx/SCFToCalyx.cpp#L1861) as well